cmake/Modules/ExportLibraries.cmake

```diff
@@ -0,0 +1,25 @@
+include(CMakePackageConfigHelpers)
+write_basic_package_Version_file("BoostConfigVersion.cmake"
+  VERSION ${BOOST_VERSION}
+  COMPATIBILITY AnyNewerVersion)
+
+get_property(BOOST_FIND_PACKAGE GLOBAL PROPERTY Boost_Find_Package)
+if(BOOST_FIND_PACKAGE)
+  # The list may be empty
+  list(REMOVE_DUPLICATES BOOST_FIND_PACKAGE)
+endif()
+configure_file("BoostConfig.cmake.in" "${CMAKE_CURRENT_BINARY_DIR}/BoostConfig.cmake" @ONLY)
+
+install(FILES
+  "${CMAKE_CURRENT_BINARY_DIR}/BoostConfig.cmake"
+  "${CMAKE_CURRENT_BINARY_DIR}/BoostConfigVersion.cmake"
+  DESTINATION lib/cmake/Boost
+)
+install(DIRECTORY ${BOOST_SOURCE}/boost
+  DESTINATION include
+)
+install(EXPORT boost-libs
```

Comment:
In order to get dependency searching working per idiomatic CMake config package methodology, you're going to have to generate an export target *per boost component*. This means, for example, having:

* BoostFilesystemTargets.cmake
* BoostThreadTargets.cmake

etc.

Reply:
I haven't found any documentation for this, not sure if that's the way to go.

Reply:
If you look at [this page](https://cmake.org/cmake/help/v3.6/manual/cmake-packages.7.html#id17) (a bit towards the bottom of that linked section), you will see an example of how they handle components in config packages:

```
include(CMakeFindDependencyMacro)
find_dependency(Stats 2.6.4)

include("${CMAKE_CURRENT_LIST_DIR}/ClimbingStatsTargets.cmake")
include("${CMAKE_CURRENT_LIST_DIR}/ClimbingStatsMacros.cmake")

set(_supported_components Plot Table)

foreach(_comp ${ClimbingStats_FIND_COMPONENTS})
  if (NOT ";${_supported_components};" MATCHES _comp)
    set(ClimbingStats_FOUND False)
    set(ClimbingStats_NOTFOUND_MESSAGE "Unsupported component: ${_comp}")
  endif()
  include("${CMAKE_CURRENT_LIST_DIR}/ClimbingStats${_comp}Targets.cmake")
endforeach()
```

This is what I'm basing my suggestion on. Would this make sense for boost?

Reply:
Hmm... That's just extra complexity for the sake of it, I don't really see the point of adding that.
It could be possible to add easily though.

Reply:
You're the boss, I'm just sharing my thoughts. Honestly I don't have much experience with configure packages so I'm not sure what is right, wrong, or ideal

Reply:
I don't have that much experience either, to be honest!

Reply:
So let me ask then: If I do this:

    find_package( boost COMPONENTS filesystem )

Will I also have targets like `Boost::thread`? If so, I think *that's* why separate target.cmake files are output, so that you only define the import targets that the user asked for. If it's all in one big file I think no matter what components they specify, you'll get all of them.

Am I correct on this? I'm having some work machine problems, but once I get past that I'll be able to test that theory for you.

Reply:
Yes, but I believe that's relevant for Find*.cmake modules that do a lot of extra work to find various components.
In this case, you have a pre-made list of libraries, so the impact is really minimal.

Reply:
Not sure if it's practical or not, I'm just making a point in terms of design: Having targets defined that you didn't explicitly ask for might go against what the CMake developers originally intended. However, they don't seem to document this. I'll ask on the mailing list, not because I'm particularly concerned about how you've chosen to do this, but more for my own understanding since lately I've been trying to learn more about config packages. I'll let you know what they say, if you're curious.

Thanks for discussing!

Reply:
Already got a response:

Hi Robert,


> On Fri, Sep 1, 2017 at 9:21 PM, Robert Dailey <rcdailey.lists@gmail.com> wrote:
One problem I thought of with the former (one big target.cmake with
all import targets in there) is that if you only ask for a subset of
components in find_package(), you will still get all of them since all
imports are defined in a single file.

In my project I have a bunch of components and do one exported target per component
exactly by the mentioned reason -- user didn't ask for others...
 
> Does this go against any design
principles?

As far as I know, there are no clear design principles :) (yet, at least nowadays) -- at least doing
a lot of CMake projects since 2009, I've never seen an explicit list of them %)
IMHO, there is a lack of "official guildelines" (or it is really hard to search for 'em)

> Assuming this really happens, are there any negative side
effects?

I could see the impact on build time only in this case... and for me the most obvious is increasing
time to process the lists (which is for some reasons really slow on Windows, at least in our
build farm which uses vargant and VirtualBox images)
(but I don't have any particular numbers, cuz never implemented the first approach)